contracts/container/contract.go

```diff
@@ -491,6 +494,67 @@ func List(owner []byte) [][]byte {
 	return list
 }
 
+// AddNextEpochNodes adds container members for the next epoch. Changes must be
+// committed with [CommitContainerListUpdate]. Without submitting, it is a no-op
+// operation. Results of the call operation can be received via [Nodes]. This
+// method must be called only when a container list is changed, otherwise
+// nothing should be done.
+// Call must be signed by the Alphabet nodes.
+func AddNextEpochNodes(cID []byte, nodes [][]byte) {
+	ctx := storage.GetContext()
+	multiaddr := common.AlphabetAddress()
+	common.CheckAlphabetWitness(multiaddr)
+
+	for _, node := range nodes {
+		pk := node[2 : 2+interop.PublicKeyCompressedLen] // public key in V2 marshaled node info
+		key := []byte{nextEpochNodesPrefix}
+		key = append(key, cID...)
+		key = append(key, pk...)
+
+		storage.Put(ctx, key[:limits.MaxStorageKeyLen], node)
+	}
+}
+
+// CommitContainerListUpdate commits container list changes made by
+// [AddNextEpochNodes] calls in advance. If no [AddNextEpochNodes] have been
+// made, it clears container list. Mades "ContainerUpdate" notification with
+// container ID after successful list change.
+// Call must be signed by the Alphabet nodes.
+func CommitContainerListUpdate(cID []byte) {
+	ctx := storage.GetContext()
+	multiaddr := common.AlphabetAddress()
+	common.CheckAlphabetWitness(multiaddr)
+
+	oldNodes := storage.Find(ctx, append([]byte{nodesPrefix}, cID...), storage.KeysOnly)
+	for iterator.Next(oldNodes) {
+		oldNode := iterator.Value(oldNodes).(string)
+		storage.Delete(ctx, oldNode)
+	}
+
+	newNodes := storage.Find(ctx, append([]byte{nextEpochNodesPrefix}, cID...), storage.None)
+	for iterator.Next(newNodes) {
+		newNode := iterator.Value(newNodes).(struct {
+			key []byte
+			val []byte
+		})
+
+		storage.Delete(ctx, newNode.key[:])
+
+		newKey := append([]byte{nodesPrefix}, newNode.key[1:]...)
+		storage.Put(ctx, newKey, newNode.val)
+	}
+
+	runtime.Notify("ContainerUpdate", cID)
```

Comment:
`NodesUpdate`? We will have mutable containers eventually and here it's not changed.